sphinx/pycode/__init__.py

```diff
@@ -63,11 +63,11 @@ def for_module(cls, modname: str) -> "ModuleAnalyzer":
             return entry
 
         try:
-            type, source = get_module_source(modname)
-            if type == 'string':
-                obj = cls.for_string(source, modname)
-            else:
-                obj = cls.for_file(source, modname)
+            filename, source = get_module_source(modname)
```

Comment:
Why are you changing interface of `get_module_source()`? I think this is incompatible change. Unfortunately, there are many hackers to touch sphinx module directly. So it would be better not to change our interface. Or we have to obey deprecation process to change it. But it seems no special reason to change it in this time.

Reply:
I change get_module_source() return type to be able to return filename and source of module together. Filename stored to Sphinx internals and used in Warnings test to show user exact warning location, source used in Analyzer to get docstrings.
Original interface can return only filename or only source, not both.
With a full search in Sphinx sources I found just two calls for get_module_source() - in pycode and in test_util and update both, so I hope change get_module_source() interface doesn`t break any code.
If you see better way how to implement this - just say. I`ll implement it.

Reply:
In our experience, 3rd party extension sometimes uses Sphinx core functions even undocumented. So it would be better to keep compatibility. But, as you said, it should return both values if possible.

I think it would be better to copy new `get_module_source()` function to `sphinx.pycode`. As you commented, it is used only from pycode. So it should be moved inside pycode module. And original `get_module_source()` will be marked as deprecated to remove it in future. How about this?

Reply:
May be make it "protected" (underscored) static function in ModuleAnalizer class to mark fact that this function used only by this class? And what to do with Tests? Move from test_util to test_pycode file?

Reply:
In my experience of OSS development, third party developers usually touches the inside of framework to implement their extensions. I think there are no reason to support such impolite extensions. But dropping old API will cause their users trouble. So I'd like to keep compatibility of APIs as possible.

On such sight, either public or protected is okay to me. I consider both are almost public APIs for hackers...
(If I have to choose one, IMO, I prefer to use public interface. I feel protection model is too much for python. It's like gentlemen's agreements. So it does not protect impolite hackers...)

>And what to do with Tests? Move from test_util to test_pycode file?

Yes, please copy and modify testcode to test_pycode.

Reply:
Done :-) Can you take a look to `sphinxcontrib-htmlhelp` pull request? It seems not-only-cp1251 problem - python `cpXXXX` codepages are not recognised by hh.exe See also https://www.iana.org/assignments/character-sets/character-sets.xhtml ,please.

Reply:
Take a look later (maybe in new year)